udao/optimization/soo/mogd.py

```diff
@@ -751,19 +771,39 @@ def objective_loss(
             raise NotImplementedError("Objective with only one bound is not supported")
         return loss
 
-    def compute_loss(self, problem: co.SOProblem, input_data: Any) -> th.Tensor:
-        obj_output = problem.objective.function(input_data)
+    def _compute_loss(
+        self, problem: co.SOProblem, input_data: Union[UdaoInput, Dict]
+    ) -> Tuple[th.Tensor, float, float, int, bool]:
+        obj_output = (
+            problem.objective.function(input_data)  # type: ignore
+            if isinstance(input_data, UdaoInput)
+            else problem.objective.function(**input_data)
+        )
         objective_loss = self.objective_loss(obj_output, problem.objective)
         constraint_loss = th.zeros_like(objective_loss, device=self.device)
 
         if problem.constraints:
             const_outputs = [
-                constraint.function(input_data) for constraint in problem.constraints
+                constraint.function(input_data)  # type: ignore
+                if isinstance(input_data, UdaoInput)
+                else constraint.function(**input_data)
+                for constraint in problem.constraints
             ]
             constraint_loss = self.constraints_loss(const_outputs, problem.constraints)
 
         loss = objective_loss + constraint_loss
-        return loss
+        sum_loss = th.sum(loss)
+        min_loss = th.min(loss).cpu().item()
+        min_loss_id = int(th.argmin(loss).cpu().item())
+        best_obj = obj_output[min_loss_id].cpu().item()
+
+        return (
```

Comment:
It's better to either return fewer elements (splitting into smaller functions) or return a dictionary to keep clarity on what's returned.
A long tuple can lead to difficulty reading and mistakes in the order of returned values